### PR TITLE
Issue #1653 Password can not be entered when console/terminal lacks PTY support

### LINKED ISSUE #1653: Password can not be entered when console/terminal lacks PTY support
**Steps to Reproduce**
- Open minishift project in IDEA IntelliJ / Eclipse
- Add Run configuration with `--iso-url` pointed to RHEL iso
- Click on Run button (Shift + F10)
- In the 'Console View', enter subscription username

See Devstudio issue as well https://issues.jboss.org/browse/JBIDE-25260

### With CMD / Powershell behavior

Password prompt message appear again on pressing enter.
**Expected**: Should repeat asking of username/password saying invalid credentials.

password_prompt_enter

### PR DESCRIPTION
**Pull request by praveenkumar**
Fixes #1653

### CODE REVIEW

File: pkg/util/input.go [hunk 1]
```diff
@@ -31,10 +31,14 @@ func ReadInputFromStdin(fieldlabel string) string {
...
 
+func IsTtySupported() bool {
+	return terminal.IsTerminal(syscall.Stdin)
+}
+
 func ReadPasswordFromStdin(fieldlabel string) string {
 	var value string
 	fmt.Printf("%s: ", fieldlabel)
-	pwinput, err := terminal.ReadPassword(int(os.Stdin.Fd()))
+	pwinput, err := terminal.ReadPassword(syscall.Stdin)
```
Comment (gbraad) [inline pkg/util/input.go]: Tested on Windows?

**Comment by LalatenduMohanty** [comment]
Looks like cross compilation on Windows failing

```
CGO_ENABLED=0 GOARCH=amd64 GOOS=windows go build -tags "containers_image_openpgp containers_image_storage_stub containers_image_ostree_stub exclude_graphdriver_devicemapper exclude_graphdriver_devicemapper exclude_graphdriver_btrfs exclude_graphdriver_overlay" -pkgdir=/home/travis/gopath/src/github.com/minishift/minishift/out/bindata --installsuffix cgo -ldflags="-X github.com/minishift/minishift/pkg/version.minishiftVersion=1.10.0 -X github.com/minishift/minishift/pkg/version.b2dIsoVersion=v1.2.0 -X github.com/minishift/minishift/pkg/version.centOsIsoVersion=v1.4.0 -X github.com/minishift/minishift/pkg/version.minikubeIsoVersion=v0.24.0 -X github.com/minishift/minishift/pkg/version.openshiftVersion=v3.6.1 -X github.com/minishift/minishift/pkg/version.commitSha=ae7a1f2 -s -w -extldflags '-static'" -o out/windows-amd64/minishift.exe ./cmd/minishift

# github.com/minishift/minishift/pkg/util

pkg/util/input.go:35: cannot use syscall.Stdin (type syscall.Handle) as type int in argument to terminal.IsTerminal

pkg/util/input.go:41: cannot use syscall.Stdin (type syscall.Handle) as type int in argument to terminal.ReadPassword

make: *** [out/windows-amd64/minishift.exe] Error 2
```

File: pkg/minishift/hostfolder/hostfolder.go [hunk 2]
```diff
@@ -141,6 +144,9 @@ func Add(name string, allInstances bool) error {
 	}
 	mountpoint := readInputForMountpoint(name)
 	username := util.ReadInputFromStdin("Username")
+	if !util.IsTtySupported() {
+		return fmt.Errorf("Not a tty supported terminal")
```
Comment (LalatenduMohanty) [inline pkg/minishift/hostfolder/hostfolder.go]: You should define an error string and then reuse the same for errors.

File: pkg/util/input.go [hunk 3]
```diff
@@ -31,10 +31,14 @@ func ReadInputFromStdin(fieldlabel string) string {
 	return value
 }
 
+func IsTtySupported() bool {
+	return terminal.IsTerminal(syscall.Stdin)
+}
+
```
Comment (LalatenduMohanty) [inline pkg/util/input.go]: This needs to be tested on  devstudio/Eclipse.

**Comment by budhrg** [comment]
screenshot from 2017-12-19 14-49-03

/cc @robstryker

**Comment by robstryker** [comment]
That looks nice to me ;)  

Can you also include a screenshot of what happens if invalid user / pass credentials are included via command line flags? 

It'd be best if we don't even ask for username if we know we can't ask for password, so maybe recognizing the lack of TTY one line earlier would be better? Why force the user to type a username when they won't be able to type a password?

File: pkg/util/input.go [hunk 4]
```diff
@@ -31,10 +31,15 @@ func ReadInputFromStdin(fieldlabel string) string {
...
 
+func IsTtySupported() bool {
+	// https://github.com/golang/go/issues/12978
+	return terminal.IsTerminal(int(syscall.Stdin))
+}
+
 func ReadPasswordFromStdin(fieldlabel string) string {
 	var value string
 	fmt.Printf("%s: ", fieldlabel)
-	pwinput, err := terminal.ReadPassword(int(os.Stdin.Fd()))
```
Comment (gbraad) [inline pkg/util/input.go]: will not work on Windows

**Comment by gbraad** [comment]
> It'd be best if we don't even ask for username if we know we can't ask for password, 

disallow in general? I am OK with this.

**Comment by robstryker** [comment]
tested on fedora, seems to work as expected except that part where it asks me for a username and then doesn't let me type a password. Better to error before asking for username if possible.  If it's not possible, then I guess this is still way better than it was before.

**Comment by gbraad** [comment]
Also consider changing the description to reflect this change

**Pull request merged**